flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecWindowTableFunction.java

```diff
@@ -103,4 +111,67 @@ protected Transformation<RowData> translateToPlanInternal(
                 inputTransform.getParallelism(),
                 false);
     }
+
+    private Transformation<RowData> translateWithUnalignedWindow(
+            PlannerBase planner,
+            ExecNodeConfig config,
+            RowType inputRowType,
+            Transformation<RowData> inputTransform) {
+        final WindowTableFunctionOperatorBase windowTableFunctionOperator =
+                createUnalignedWindowTableFunctionOperator(config, inputRowType);
+        final OneInputTransformation<RowData, RowData> transform =
+                ExecNodeUtil.createOneInputTransformation(
+                        inputTransform,
+                        createTransformationMeta(WINDOW_TRANSFORMATION, config),
+                        windowTableFunctionOperator,
+                        InternalTypeInfo.of(getOutputType()),
+                        inputTransform.getParallelism(),
+                        false);
+
+        final int[] partitionKeys = extractPartitionKeys(windowingStrategy.getWindow());
```

Comment:
If user doesn't specify the partition key in session tvf, do we need to set the state KeySelector？ The session window tvf operator is stateful,  I want to say is that if user doesn't specify the key, the session operator is singleton, so what is the state key? Moreover, batch mode doesn't need to set KeySelector.

Reply:
If user doesn't specify the partition keys, a EmptyRowDataKeySelector will be returned.
I'll split this function into StreamXXX and BatchXXX.